File: src/nodes/vocab/kinds.py
from __future__ import annotations

from nodes.kernel.registry import KindSpec, Registry
from nodes.vocab.source import SOURCE, require_identifiable_source

NOTE = "note"
IDEA = "idea"
QUESTION = "question"
TOPIC = "topic"
PAPER = "paper"
BOOK = "book"
DATASET = "dataset"

PROSE_KINDS = (NOTE, IDEA, QUESTION, TOPIC)
SOURCE_KINDS = (PAPER, BOOK, DATASET)


def register_knowledge_vocab(reg: Registry) -> None:
    """Register the standard knowledge-vocab kinds onto `reg`.

    Mirrors `nodes.kernel.shapes.register_builtin_shapes`.
    """
    for name in PROSE_KINDS:
        reg.register(KindSpec(name=name))
    for name in SOURCE_KINDS:
        reg.register(
            KindSpec(
                name=name,
                required_facets={SOURCE},
                invariants=[require_identifiable_source],
            )
        )

File: ts/src/vocab/kinds.ts
import type { Registry } from "../registry.js";
import { SOURCE, requireIdentifiableSource } from "./source.js";

export const NOTE = "note";
export const IDEA = "idea";
export const QUESTION = "question";
export const TOPIC = "topic";
export const PAPER = "paper";
export const BOOK = "book";
export const DATASET = "dataset";

export const PROSE_KINDS = [NOTE, IDEA, QUESTION, TOPIC] as const;
export const SOURCE_KINDS = [PAPER, BOOK, DATASET] as const;

/** Register the standard knowledge-vocab kinds onto `reg`.
 *  Mirrors `registerBuiltinShapes` in `shapes.ts`. */
export function registerKnowledgeVocab(reg: Registry): void {
  for (const name of PROSE_KINDS) {
    reg.register({ name });
  }
  for (const name of SOURCE_KINDS) {
    reg.register({ name, requiredFacets: new Set([SOURCE]), invariants: [requireIdentifiableSource] });
  }
}
